File: LeetCode/Hao/33SearchInRotatedSortedArray.java
public class Solution {
    public int search(int[] nums, int target) {
        int left, right;
        
        left = 0;
        right = nums.length - 1;
        
        while (left <= right) {
            if (left == right) {
                break;
            }
            
            int mid = left + (right - left) / 2;
            
            if (target == nums[mid]) {
                left = mid;
                break;
            } else if (target == nums[0]) {
                left = 0;
                break;
            } else if (target < nums[0] && nums[mid] >= nums[0]) {
                left = mid + 1;
            } else if (target < nums[0] && target < nums[mid]) {
                right = mid;
            } else if (target < nums[0] && target > nums[mid]) {
                left = mid + 1;
            } else if (target > nums[0] && nums[mid] < nums[0]) {
                right = mid;
            } else if (target > nums[0] && target > nums[mid]) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        
        if (nums[left] == target) {
            return left;
        }
        
        return -1;
    }
}

File: LeetCode/Liang/33SearchInRotatedSortedArray.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        if (nums.empty()) {
            return -1;
        }
        
        int begin = 0, end = nums.size() - 1;
        while (begin < end) {
            int mid = begin + (end - begin) / 2;
            if (nums.at(mid) > nums.at(end)) {
                if (target <= nums.at(mid) && target > nums.at(end)) {
                    end = mid;
                } else {
                    begin = mid + 1;
                }
            } else {
                if (target <= nums.at(end) && target > nums.at(mid)) {
                    begin = mid + 1;
                } else {
                    end = mid;
                }
            }
        }
        
        if (nums.at(begin) == target) {
            return begin;
        } else {
            return -1;
        }
    }
};
